utils/testgen.py

```diff
@@ -328,92 +259,53 @@ def pytest_generate_tests(metafunc):
         for more details.
 
     """
+    pf = ProviderFilter(types=types, required_fields=required_fields)
+    return providers(metafunc, filters=[pf])
 
-    argnames = []
-    argvalues = []
-    idlist = []
-
-    for provider in cfme_data.get('management_systems', {}):
-
-        # Check provider hasn't been filtered out with --use-provider
-        if provider not in filtered:
-            continue
 
-        try:
-            prov_obj = get_crud(provider)
-        except UnknownProviderType:
-            continue
-
-        if not prov_obj:
-            logger.debug("Whilst trying to create an object for %s we failed", provider)
-            continue
-
-        if provider_types is not None:
-            if not(prov_obj.type_tclass in provider_types or prov_obj.type_name in provider_types):
-                continue
-
-        # Run through all the testgen uncollect fns
-        uncollect = False
-        uncollect_fns = [_uncollect_restricted_version, _check_required_fields,
```

Comment:
Though perhaps unwieldy this section of code was the one that supplanted the need for that suprt long conditional above.